### PR TITLE
[Inductor] Autotuner Integration

### PR DESCRIPTION
**Pull request by spectrometerHBH**
```sh
# E2E Integration
## XGB BASELINE
rm -rf /scratch/bohanhou/fresh/hf-model-0-0-training/;TORCHINDUCTOR_AUTOTUNER_PATH=/scratch/bohanhou/fresh/new_experiments/xgb_baseline/xgb_baseline.pkl TORCHINDUCTOR_CACHE_DIR=/scratch/bohanhou/fresh/hf-model-0-0-training/ TORCHINDUCTOR_AUTOTUNER_SPACE=0 TORCH_LOGS="+inductor" TORCHINDUCTOR_BENCHMARK_KERNEL=1 python3 benchmarks/dynamo/huggingface.py --amp --performance --training --inductor -d cuda --device-index 0 --output hf-model-0-0-training.csv >./hf-model-0-0-training.log 2>&1
rm -rf /scratch/bohanhou/fresh/hf-model-0-2-training/;TORCHINDUCTOR_AUTOTUNER_PATH=/scratch/bohanhou/fresh/new_experiments/xgb_baseline/xgb_baseline.pkl TORCHINDUCTOR_CACHE_DIR=/scratch/bohanhou/fresh/hf-model-0-2-training/ TORCHINDUCTOR_AUTOTUNER_SPACE=2 TORCH_LOGS="+inductor" TORCHINDUCTOR_BENCHMARK_KERNEL=1 python3 benchmarks/dynamo/huggingface.py --amp --performance --training --inductor -d cuda --device-index 0 --output hf-model-0-2-training.csv >./hf-model-0-2-training.log 2>&1
## NN Pointwise
rm -rf /scratch/bohanhou/fresh/hf-model-1-0-training/;TORCHINDUCTOR_AUTOTUNER_PATH=/scratch/bohanhou/fresh/new_experiments/nn_pointwise/nn_pointwise_True_1400_0.0420697318410453_0.02314407822228018.pkl TORCHINDUCTOR_CACHE_DIR=/scratch/bohanhou/fresh/hf-model-1-0-training/ TORCHINDUCTOR_AUTOTUNER_SPACE=0 TORCH_LOGS="+inductor" TORCHINDUCTOR_BENCHMARK_KERNEL=1 python3 benchmarks/dynamo/huggingface.py --amp --performance --training --inductor -d cuda --device-index 0 --output hf-model-1-0-training.csv >./hf-model-1-0-training.log 2>&1
rm -rf /scratch/bohanhou/fresh/hf-model-1-2-training/;TORCHINDUCTOR_AUTOTUNER_PATH=/scratch/bohanhou/fresh/new_experiments/nn_pointwise/nn_pointwise_True_1400_0.0420697318410453_0.02314407822228018.pkl TORCHINDUCTOR_CACHE_DIR=/scratch/bohanhou/fresh/hf-model-1-2-training/ TORCHINDUCTOR_AUTOTUNER_SPACE=2 TORCH_LOGS="+inductor" TORCHINDUCTOR_BENCHMARK_KERNEL=1 python3 benchmarks/dynamo/huggingface.py --amp --performance --training --inductor -d cuda --device-index 0 --output hf-model-1-2-training.csv >./hf-model-1-2-training.log 2>&1
## NN L2R
rm -rf /scratch/bohanhou/fresh/hf-model-2-0-training/;TORCHINDUCTOR_AUTOTUNER_PATH=/scratch/bohanhou/fresh/new_experiments/nn_l2r/nn_l2r_True_700_0.0003389147304638488_0.8339776992797852.pkl TORCHINDUCTOR_CACHE_DIR=/scratch/bohanhou/fresh/hf-model-2-0-training/ TORCHINDUCTOR_AUTOTUNER_SPACE=0 TORCH_LOGS="+inductor" TORCHINDUCTOR_BENCHMARK_KERNEL=1 python3 benchmarks/dynamo/huggingface.py --amp --performance --training --inductor -d cuda --device-index 0 --output hf-model-2-0-training.csv >./hf-model-2-0-training.log 2>&1
```

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #107615
* #107489
* __->__ #107488

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @Xia-Weiwen @ngimel

### CODE REVIEW

File: torch/_inductor/autotuner/model.py
Comment (jansel) [inline torch/_inductor/autotuner/model.py:1]: This file needs comments and tests

File: torch/_inductor/autotuner/model.py [hunk 1]
```diff
@@ -0,0 +1,801 @@
...
+    ]
+    read_write: dependencies.ReadWrites
+    src_code: str
+    autotuner_dict: Dict
+
+    def __repr__(self):
+        return f"AutotunerRawData({self.io_deps}, {self.read_write}, {self.src_code}, {self.autotuner_dict})"
+
+
+def get_reads_writes(cur_scheduler, node, src_code):
```
Comment (jansel) [inline torch/_inductor/autotuner/model.py:43]: Would these be better as methods on the scheduler objects?

File: torch/_inductor/autotuner/model.py [hunk 2]
```diff
@@ -0,0 +1,801 @@
...
+        strides,
+        sizes,
+        node_bytes,
+    )
+
+
+### feature extraction related
+
+# op_dict needs to be deterministic
+OP_DICT = {
```
Comment (jansel) [inline torch/_inductor/autotuner/model.py:107]: Need a way to keep this updated and/or check that is up to date

File: torch/_inductor/autotuner/model.py [hunk 3]
```diff
@@ -0,0 +1,801 @@
...
+    num_stages: int
+
+
+class KernelCategory(enumerate):
+    POINTWISE = 0
+    REDUCTION = 1
+    PERSISTENT_REDUCTION = 2
+
+
+def get_kernel_category(src: str) -> KernelCategory:
```
Comment (jansel) [inline torch/_inductor/autotuner/model.py:205]: same comments from other version of PR, need testing

File: torch/_inductor/autotuner/model.py [hunk 4]
```diff
@@ -0,0 +1,801 @@
...
+
+
+class ModelType(enumerate):
+    XGB_BASELINE = 0
+    NN_POINTWISE = 1
+    NN_PAIRWISE = 2
+    NN_PAIRWISE_SMALL = 3
+
+
+class Autotuner_FFN(nn.Module):
```
Comment (jansel) [inline torch/_inductor/autotuner/model.py:313]: No underscores in class names.

File: torch/_inductor/autotuner/model.py [hunk 5]
```diff
@@ -0,0 +1,801 @@
...
+                        np.log2(np.abs(dep_feature.sizes) + 1)
+                        * np.sign(dep_feature.sizes)
+                    )
+                    + [
+                        dep_feature.is_contiguous,
+                        dep_feature.is_scalar,
+                        dep_feature.is_indirect,
+                    ]
+                )
+            feature_vecs = torch.from_numpy(np.array(feature_vecs)).to("cuda")
```
Comment (jansel) [inline torch/_inductor/autotuner/model.py:411]: torch.tensor() rather than numpy.array

File: torch/_inductor/autotuner/model.py [hunk 6]
```diff
@@ -0,0 +1,801 @@
...
+            torch.nn.init.xavier_normal_(layer.weight)
+            torch.nn.init.zeros_(layer.bias)
+
+        self.activation = model_cfg["activation"]
+
+    def forward(self, x):
+        if not isinstance(x, list):
+            x = [x]
+        x_kernel_category = np.array([i.kernel_feature.kernel_category for i in x])
+        x_kernel_category = torch.from_numpy(x_kernel_category).to("cuda").long()
```
Comment (jansel) [inline torch/_inductor/autotuner/model.py:371]:
```suggestion
        x_kernel_category = torch.tensor(x_kernel_category, device="cuda")
```

File: torch/_inductor/autotuner/model.py [hunk 7]
```diff
@@ -0,0 +1,801 @@
...
+        x_kernel_category = self.kernel_category_embedding(x_kernel_category)
+
+        x_num_of_loops = np.array([i.kernel_feature.num_of_loops for i in x])
+        x_num_of_loops = torch.from_numpy(x_num_of_loops).to("cuda").long()
```
Comment (jansel) [inline torch/_inductor/autotuner/model.py:375]: Throughout the file we are hardcoding cuda, we should make this work with other devices (for example of cuda isn't available).

File: torch/_inductor/autotuner/model.py [hunk 8]
```diff
@@ -0,0 +1,801 @@
...
+        )
+
+        torch.nn.init.xavier_normal_(self.kernel_category_embedding.weight)
+        torch.nn.init.xavier_normal_(self.num_of_loops_embedding.weight)
+        torch.nn.init.xavier_normal_(self.is_contiguous_ln.weight)
+        torch.nn.init.xavier_normal_(self.is_scalar_ln.weight)
+        torch.nn.init.xavier_normal_(self.is_indirect_ln.weight)
+        for layer in list(self.op_bag_ln) + list(self.layers):
+            torch.nn.init.xavier_normal_(layer.weight)
+            torch.nn.init.zeros_(layer.bias)
```
Comment (jansel) [inline torch/_inductor/autotuner/model.py:363]: is this stuff needed?

File: torch/_inductor/autotuner/model.py [hunk 9]
```diff
@@ -0,0 +1,801 @@
...
+            [
+                self.op_bag_ln[i](x_op_vec[:, i].unsqueeze(1))
+                for i in range(len(OP_DICT))
+            ],
+            dim=1,
+        )
+
+        x_size_hints = np.array([i.kernel_feature.size_hints for i in x])
+        x_size_hints = np.log2(x_size_hints + 1)
+        x_size_hints = torch.from_numpy(x_size_hints).to("cuda")
```
Comment (jansel) [inline torch/_inductor/autotuner/model.py:390]: use pytorch ops not numpy (throughout file)

File: torch/_inductor/autotuner/model.py [hunk 10]
```diff
@@ -0,0 +1,801 @@
...
+
+        rest_vec = np.array(
+            [
+                [i.XBLOCK, i.YBLOCK, i.RBLOCK, i.num_warps, i.num_stages]
+                + list(i.kernel_feature.numels)
+                for i in x
+            ]
+        )
+        rest_vec[:, 0:3] = np.log2(rest_vec[:, 0:3] + 1)
+        rest_vec[:, -3:] = np.log2(rest_vec[:, -3:] + 1)
```
Comment (jansel) [inline torch/_inductor/autotuner/model.py:439]: hardcoded indexing? perhaps just apply log2 above

**Review by jansel — CHANGES_REQUESTED** [review]
Also address remaining comments from prior PR

File: torch/_inductor/autotuner/model.py (hunk 8 above)
Comment (spectrometerHBH) [reply torch/_inductor/autotuner/model.py]: removed

File: torch/_inductor/autotuner/model.py (hunk 2 above)
Comment (spectrometerHBH) [reply torch/_inductor/autotuner/model.py:109]: Currently it will throw if it encounters unknown op:
https://github.com/pytorch/pytorch/pull/107488/files#diff-351bdd9a44612eb5097abb55e76cdfd58f1871cf1004ce6aae82a0077b2194f5R751

Likely a way to notify people to update.

File: torch/_inductor/autotuner/model.py (hunk 1 above)
Comment (spectrometerHBH) [reply torch/_inductor/autotuner/model.py:46]: It's only used here so I feel for now it's great to put it in this file.

File: torch/_inductor/autotuner/model.py [hunk 11]
```diff
@@ -0,0 +1,833 @@
...
+    def feature_vector_to_xgb_input(self, feature_vecs):
+        xgb_input = list()
+        for feature_vector in feature_vecs:
+            xgb_input.append(
+                [
+                    feature_vector.kernel_feature.kernel_category,
+                    feature_vector.kernel_feature.num_of_loops,
+                ]
+                + list(feature_vector.kernel_feature.op_vec)
+                + list(feature_vector.kernel_feature.size_hints)
```
Comment (jansel) [inline torch/_inductor/autotuner/model.py:697]:
```suggestion
                [
                    feature_vector.kernel_feature.kernel_category,
                    feature_vector.kernel_feature.num_of_loops,
                    *feature_vector.kernel_feature.op_vec,
                    *feature_vector.kernel_feature.size_hints,
```

and same below

**Review by jansel — CHANGES_REQUESTED** [review]

**Comment by jansel** [comment]
Still needs tests + doctstrings
